packages/dev/core/test/unit/DeviceInput/babylon.deviceInput.test.ts

```diff
@@ -0,0 +1,226 @@
+import type { Nullable } from "core/types";
+
+/**
+ * This class mocks up the basic functionality required from the DeviceInputSystem
+ */
+class MockDeviceInputSystem implements IDeviceInputSystem {
+    constructor(
+        public onDeviceConnected: Nullable<(deviceType: DeviceType, deviceSlot: number) => void>,
+        public onDeviceDisconnected: Nullable<(deviceType: DeviceType, deviceSlot: number) => void>,
+        public onInputChanged: Nullable<(deviceType: DeviceType, deviceSlot: number, eventData: IUIEvent) => void>
+    ) {}
+    pollInput(deviceType: DeviceType, deviceSlot: number, inputIndex: number): number {
+        if (deviceType === DeviceType.Mouse && deviceSlot === 0 && inputIndex === PointerInput.LeftClick) {
+            return 1;
+        } else {
+            return 0;
+        }
+    }
+    isDeviceAvailable(deviceType: DeviceType): boolean {
+        return deviceType === DeviceType.Mouse || deviceType === DeviceType.Touch;
+    }
+    dispose(): void {
+        // Do nothing
+    }
+}
+
+class MockInternalDeviceSourceManager {
+    public _refCount = 0;
+
+    constructor() {}
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    public readonly registerManager = (manager: IObservableManager): void => {
+        // Do nothing
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    public readonly unregisterManager = (manager: IObservableManager): void => {
+        // Do nothing
+    };
+
+    public dispose(): void {}
+}
+
+describe("DeviceSource", () => {
+    it("should exist", () => {
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        expect(mouseSource).not.toBeNull();
+        expect(mouseSource.deviceType).toBe(DeviceType.Mouse);
+        expect(mouseSource.deviceSlot).toBe(0);
+        expect(mouseSource.onInputChangedObservable).not.toBeUndefined();
+    });
+    it("can poll with getInput", () => {
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        const leftClick = mouseSource.getInput(PointerInput.LeftClick);
+        expect(leftClick).toBe(1);
+    });
+    it("can use onInputChangedObservable", () => {
+        expect.assertions(2);
+        let resultEvent: Nullable<IWheelEvent | IPointerEvent> = null;
+        let onInputChanged = null;
+        // Create DeviceSource and add to its observable
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, onInputChanged);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        onInputChanged = (deviceType: DeviceType, deviceSlot: number, eventData: IUIEvent) => {
+            mouseSource.onInputChangedObservable.notifyObservers(eventData as IPointerEvent);
+        };

```

Comment:
I'm not sure this is doing. I think it is simulating a mouse event, but I would expect this to be in the mock class with a public method to simulate input things (e.g., clicking a mouse button, etc.)

Reply:
I've added a function to create an event and send it through the input system.  It should move all the way throught he onInputChanged pipeline.